Update MSC4311 tests to conform to the spec and better coverage

turt2live opened the pull request:
Update MSC4311 tests to conform to the spec. MSC4311 states that the client API should still use the stripped state event format. For the federation API, it should use full PDU's. MSC4311 also requires the `m.room.create` event be included in the list of stripped state.

["Stripped state"](https://spec.matrix.org/v1.18/client-server-api/#stripped-state) is a generic term that refers to simplified slice of state shared in `invite_room_state`/`knock_room_state` (`invite_state`/`knock_state` with `/sync`) before someone is joined to a room and can see the full room state.

This PR was originally authored by @turt2live taking a stab at fixing single the flawed test (self-reported: ~~"I have no idea what I'm doing" disclaimer goes here~~) but has since been taken over by @MadLittleMods. And has naturally evolved some new better test coverage as I've worked on a full solution.

Synapse PR: ~~https://github.com/element-hq/synapse/pull/18822~~ https://github.com/element-hq/synapse/pull/19723

The previous tests passed because of a flawed implementation in Synapse which is being removed in https://github.com/element-hq/synapse/pull/19723 as well.




### Dev notes

Running the tests with [Synapse](https://github.com/element-hq/synapse):

```
COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh -run TestMSC4311StrippedStateClientAPI

COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh -run TestMSC4311FullEventsOnStrippedStateFederation

COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh -run TestMSC4311RejectInvalidStrippedStateFederation
```

```
go test ./match -v -count=1 -run TestJSONArraySome
```

turt2live committed:
Try to modify MSC4311 test to meet new proposal

kegsay commented on tests/v12_test.go:
```diff
@@ -1357,39 +1357,54 @@ func asEventIDs(pdus []gomatrixserverlib.PDU) []string {
-				must.MatchGJSON(t, ev,
-					match.JSONKeyPresent("origin_server_ts"),
-				)
+	srv.Mux().HandleFunc("/_matrix/federation/v2/invite/{roomID}/{eventID}", srv.ValidFederationRequest(t, func(fr *fclient.FederationRequest, pathParams map[string]string) util.JSONResponse {
```
You should be able to alternatively use:
```go
	srv := federation.NewServer(t, deployment,
		federation.HandleInviteRequests(func(p gomatrixserverlib.PDU) {
			// checks here
		}),
		federation.HandleKeyRequests(),
		federation.HandleMakeSendJoinRequests(),
		federation.HandleTransactionRequests(nil, nil),
	)
```

MadLittleMods replied:
I went down this route but the problem is that [`federation.HandleInviteRequests(...)`](https://github.com/matrix-org/complement/blob/bb8d0f02565254357ddfcacf629c79476b2ad9e1/federation/handle.go#L202-L206) doesn't allow you to inspect the `invite_room_state` yet. It would have to be updated to pass in the full invite request to the callback.

And `gomatrixserverlib`  [needs to be updated](https://github.com/matrix-org/gomatrixserverlib/blob/20c9de33969e2df883eaa69c283d6a38efce9284/fclient/invitev2.go#L33) to handle the new world where `invite_room_state` can be stripped state or full PDUs.

We could have an infallible `InviteV2Request.StrippedInviteRoomState()` that would give a view of stripped state regardless of whether we were passed stripped or full PDU's (we can derive stripped state from the full PDU's). And then a fallible `InviteV2Request.FullInviteRoomState()` that would return an error if the homeserver didn't pass us full PDUs.

We would also need to update [`IRoomVersion`](https://github.com/matrix-org/gomatrixserverlib/blob/20c9de33969e2df883eaa69c283d6a38efce9284/eventversion.go#L17-L42) to add new fields like `CreateEventRequiredInInviteRoomState` and ~~`FullPDUInviteRoomState`~~ (full PDU's can happen in any room version) so we can conditionally apply this behavior. Or maybe the flags could be combined as `MSC4311InviteRoomState`

> If any of the events are not a PDU, not for the room ID specified, or fail [signature checks](https://spec.matrix.org/v1.15/server-server-api/#validating-hashes-and-signatures-on-received-events), or the `m.room.create` event is missing, the receiving server MAY respond to invites with a `400 M_MISSING_PARAM` standard Matrix error (new to the endpoint). For invites to room version 12+ rooms, servers SHOULD rather than MAY respond to such requests with `400 M_MISSING_PARAM`.
>
> *-- https://github.com/matrix-org/matrix-spec-proposals/pull/4311*

kegsay committed:
Merge branch 'main' into travis/msc4311-v2

MadLittleMods committed:
Merge branch 'main' into travis/msc4311-v2

MadLittleMods committed:
More robust test

MadLittleMods commented on tests/v12_test.go:
```diff
@@ -1338,39 +1339,117 @@ func asEventIDs(pdus []gomatrixserverlib.PDU) []string {
+				// Each event should have extra fields `origin_server_ts` that indicate we're
+				// seeing a full PDU and not just a "stripped state event"
+				return should.MatchGJSON(event, match.JSONKeyPresent("origin_server_ts"))
+			}),
```
This currently fails in Synapse because it just pulls `invite_room_state` from the the `unsigned` part of the PDU which will be the stripped state client version, see [`synapse/federation/federation_client.py#L1361`](https://github.com/element-hq/synapse/blob/6100f6e4f7fb0c72f1ae2802683ebc811c0e3a77/synapse/federation/federation_client.py#L1361)

<details>
<summary>Example flawed <code>/_matrix/federation/v2/invite/{roomID}/{eventID}</code> request body from Synapse</summary>

```json
{
  "event": {
    "auth_events": [
      "$3-FVPb1HO_ZyxolbnthZ8qTVEMXyZ6px4qUdrL0exII",
      "$hRMnmEtlMElF2pqbd23ccVSfAIUq-Dqo9kxORAbs2Pk",
      "$luBrxrHgZcSqpdm5UMAXtGKxCvmOwGAHJrZ0iMeWrdw"
    ],
    "content": {
      "membership": "invite"
    },
    "depth": 6,
    "hashes": {
      "sha256": "xBOaaPtS1jAtu/qsWiBIXFZqDdOm6zI8R+RVgXfhm00"
    },
    "origin_server_ts": 1777676386564,
    "prev_events": [
      "$VoTAQZqsV4zWnRi3Uh9n42ioJ4J74UvwMoZuHNiHWU4"
    ],
    "room_id": "!gvlLh8tsK9LSCdF8y56dVfErfUiw7lZnLQ8JTB_kih4",
    "sender": "@user-1-alice:hs1",
    "signatures": {
      "hs1": {
        "ed25519:a_NKNg": "xS/15iTHLn3I65oaTdVRMFBrs+ASbKp9Lm9GM8oLoKlD3ZUwtuBP2pyN/tdUXgraKnRNDU1b3PLxpXYzsLGkCQ"
      }
    },
    "state_key": "@bob:host.docker.internal:44453",
    "type": "m.room.member",
    "unsigned": {
      "age": 11,
      "invite_room_state": [
        {
          "auth_events": [],
          "content": {
            "room_version": "12"
          },
          "depth": 1,
          "hashes": {
            "sha256": "vVkUaUxHUum571Kj4D0vo85iS2dssrPpILSm3XnELpc"
          },
          "origin_server_ts": 1777676385521,
          "prev_events": [],
          "sender": "@user-1-alice:hs1",
          "signatures": {
            "hs1": {
              "ed25519:a_NKNg": "m7pf3WIpaGZEgTR85+L+ewSY88eBXJSGecOXXJmLudEFq4mGxR4MBw6/x9+xMTz4byo+z0a+0h4IcvvSyYhiAA"
            }
          },
          "state_key": "",
          "type": "m.room.create",
          "unsigned": {
            "age_ts": 1777676385521
          }
        },
        {
          "content": {
            "join_rule": "public"
          },
          "sender": "@user-1-alice:hs1",
          "state_key": "",
          "type": "m.room.join_rules"
        },
        {
          "content": {
            "displayname": "user-1-alice",
            "membership": "join"
          },
          "sender": "@user-1-alice:hs1",
          "state_key": "@user-1-alice:hs1",
          "type": "m.room.member"
        }
      ]
    }
  },
  "invite_room_state": [
    {
      "auth_events": [],
      "content": {
        "room_version": "12"
      },
      "depth": 1,
      "hashes": {
        "sha256": "vVkUaUxHUum571Kj4D0vo85iS2dssrPpILSm3XnELpc"
      },
      "origin_server_ts": 1777676385521,
      "prev_events": [],
      "sender": "@user-1-alice:hs1",
      "signatures": {
        "hs1": {
          "ed25519:a_NKNg": "m7pf3WIpaGZEgTR85+L+ewSY88eBXJSGecOXXJmLudEFq4mGxR4MBw6/x9+xMTz4byo+z0a+0h4IcvvSyYhiAA"
        }
      },
      "state_key": "",
      "type": "m.room.create",
      "unsigned": {
        "age_ts": 1777676385521
      }
    },
    {
      "content": {
        "join_rule": "public"
      },
      "sender": "@user-1-alice:hs1",
      "state_key": "",
      "type": "m.room.join_rules"
    },
    {
      "content": {
        "displayname": "user-1-alice",
        "membership": "join"
      },
      "sender": "@user-1-alice:hs1",
      "state_key": "@user-1-alice:hs1",
      "type": "m.room.member"
    }
  ],
  "room_version": "12"
}
```

</details>

MadLittleMods replied:
Test passes with Synapse with the changes from https://github.com/element-hq/synapse/pull/19723 :white_check_mark: 

```
COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh -run TestMSC4311FullEventsOnStrippedStateFederation
```

MadLittleMods commented on tests/v12_test.go:
```diff
@@ -1338,39 +1339,117 @@ func asEventIDs(pdus []gomatrixserverlib.PDU) []string {
+// iterating stops and we return.
+//
+// Will fail if the array is empty and the check never runs
+func JSONArraySome(wantKey string, fn func(gjson.Result) error) match.JSON {
```
Opinions on whether this is suitable to add to [`match/json.go`](https://github.com/matrix-org/complement/blob/bb8d0f02565254357ddfcacf629c79476b2ad9e1/match/json.go). Seems general and good enough to me.

MadLittleMods replied:
Moved as an official matcher :fast_forward:

MadLittleMods committed:
Add TODO about testing `knock_room_state`

MadLittleMods committed:
Fix raw bytes being encoded as `event` instead of actual JSON

MadLittleMods commented on tests/v12_test.go:
```diff
@@ -1338,39 +1339,133 @@ func asEventIDs(pdus []gomatrixserverlib.PDU) []string {
+
+		var roomVersion gomatrixserverlib.RoomVersion
+		if err := json.Unmarshal([]byte(rawRoomVersion), &roomVersion); err != nil {
+			t.Fatalf("failed to parse room version: %s", err)
```
Better way to get a `gomatrixserverlib.RoomVersion` from a room version string?

MadLittleMods commented on tests/v12_test.go:
```diff
@@ -1338,39 +1339,133 @@ func asEventIDs(pdus []gomatrixserverlib.PDU) []string {
-func TestMSC4311FullCreateEventOnStrippedState(t *testing.T) {
+// Alice will invite Bob. Bob's server should receive full PDUs in `invite_room_state`
+// according to MSC4311
+func TestMSC4311FullEventsOnStrippedStateFederation(t *testing.T) {
```
TODO: Add test for the client API where the `m.room.create` event is required but all they are all still stripped state.

MadLittleMods replied:
Added `TestMSC4311StrippedStateClientApi` ✅

MadLittleMods committed:
Use more specific type

MadLittleMods committed:
WIP: Add `TestMSC4311CreateEventInStrippedStateClientApi`

MadLittleMods committed:
Use more robust `MustSyncUntil`

MadLittleMods committed:
WIP: Add test for `invite_room_state` in client API

But it doesn't seem possible to see. For example,
Synapse strips it out when trying to view events,
https://github.com/element-hq/synapse/blob/6100f6e4f7fb0c72f1ae2802683ebc811c0e3a77/synapse/events/utils.py#L590-L596

MadLittleMods committed:
Remove `invite_room_state` client test

MadLittleMods committed:
Remove stray parallel

MadLittleMods committed:
Better casing

MadLittleMods committed:
Merge branch 'main' into travis/msc4311-v2

MadLittleMods committed:
Add client `knock_state` test

MadLittleMods committed:
Add `knock_room_state` test (`MustKnockRoom` not implemented)

MadLittleMods committed:
Implement `srv.MustKnockRoom`

MadLittleMods committed:
We're doing the TODO now

MadLittleMods committed:
Split up tests

MadLittleMods committed:
Add `TestMSC4311RejectInvalidStrippedStateFederation`

MadLittleMods commented on tests/v12_test.go:
```diff
@@ -1338,39 +1340,366 @@ func asEventIDs(pdus []gomatrixserverlib.PDU) []string {
+	// Alice's server should reject the invite request because of the (the tested reason)
+	// and should respond with a `400 M_MISSING_PARAM` response.
+	t.Run("parallel", func(t *testing.T) {
+		// TODO: Test events not full PDU's
+
+		// TODO: Test events not from the same room
+
+		// TODO: Test invalid signatures/hashes
```
Something for future PRs

MadLittleMods commented on tests/v12_test.go:
```diff
@@ -1338,39 +1340,366 @@ func asEventIDs(pdus []gomatrixserverlib.PDU) []string {
+// > error (new to the endpoint). For invites to room version 12+ rooms, servers
+// > SHOULD rather than MAY respond to such requests with `400 M_MISSING_PARAM`.
+func TestMSC4311RejectInvalidStrippedStateFederation(t *testing.T) {
+	runtime.SkipIf(t, runtime.Synapse)  // FIXME: Run these tests after 2027-06-01
```
Also cross-linked this in the Synapse codebase (see https://github.com/element-hq/synapse/pull/19723)

MadLittleMods committed:
Add note about this applying to other room versions

anoadragon453 commented on client/client.go:
```diff
@@ -285,6 +285,39 @@ func (c *CSAPI) InviteRoom(t ct.TestLike, roomID string, userID string) *http.Re
+//
+// Args:
+//   - `serverNames`: The list of servers to attempt to knock on the room through.
+//     These should be a resolvable addresses within the deployment network.
```
```suggestion
//     These should be one or more resolvable addresses within the deployment network.
```

MadLittleMods replied:
This is the standard copy-pasta we have on other places. I'd actually fix it this other way:

```suggestion
//     These should be a resolvable address within the deployment network.
```

I've updated our existing spots in https://github.com/matrix-org/complement/pull/890 and updated the new spots in this PR :white_check_mark:

anoadragon453 commented on tests/v12_test.go:
```diff
@@ -1338,39 +1340,374 @@ func asEventIDs(pdus []gomatrixserverlib.PDU) []string {
 
-func TestMSC4311FullCreateEventOnStrippedState(t *testing.T) {
+// MSC4311 mandates that `m.room.create` is a required event in
+// `invite_state`/`knock_state` (stripped state) in `/sync responses. MSC4311 applies
```
```suggestion
// `invite_state`/`knock_state` (stripped state) in `/sync` responses. MSC4311 applies
```

anoadragon453 commented on tests/v12_test.go:
```diff
@@ -1338,39 +1340,374 @@ func asEventIDs(pdus []gomatrixserverlib.PDU) []string {
+					string(fr.Content()),
+				)
+
+				// Invites for an unexpected rooms is an error
```
```suggestion
				// Invites for an unexpected room is an error
```

anoadragon453 commented on tests/v12_test.go:
```diff
@@ -1338,39 +1340,374 @@ func asEventIDs(pdus []gomatrixserverlib.PDU) []string {
+				// behavior we're trying to test)
+				inviteResponse := gjson.ParseBytes(fr.Content())
+				must.MatchGJSON(t, inviteResponse,
+					match.JSONArraySome("invite_room_state", func(event gjson.Result) error {
+						// MSC4311 also mandates that `m.room.create` event is required
+						return should.MatchGJSON(event, match.JSONKeyEqual("type", "m.room.create"))
+					}),
+					match.JSONArrayEach("invite_room_state", func(event gjson.Result) error {
+						// Each event should have extra fields `origin_server_ts` that indicate we're
+						// seeing a full PDU and not just a "stripped state event"
+						return should.MatchGJSON(event, match.JSONKeyPresent("origin_server_ts"))
```
Seems like it'd make sense to extract this to a helper function at this point, given I've seen it about 4 times now.

MadLittleMods replied:
I'd rather not (not totally convinced yet). They're all dealing with different field names which of course could be a function arg but the abstraction is just going to make things less clear what we care about.

anoadragon453 commented on tests/v12_test.go:
```diff
@@ -1338,39 +1340,374 @@ func asEventIDs(pdus []gomatrixserverlib.PDU) []string {
+// MSC4311 applies retroactively to any room versions but we're testing room version 12
+// as it *SHOULD* reject instead of *MAY*.
+func TestMSC4311RejectInvalidStrippedStateFederation(t *testing.T) {
+	runtime.SkipIf(t, runtime.Synapse)  // FIXME: Run these tests after 2027-06-01
```
Could you also add that this date was chosen to allow the ecosystem time to implement MSC4311? Otherwise it seems arbitrary (without having seen the Synapse code).

May also be worth adding a to-discuss issue/reminder about this :)

MadLittleMods replied:
Added the context to the comment and created https://github.com/element-hq/synapse/issues/19943 to track the task which I've cross linked in the comments and added to the To-Discuss board for next year :fast_forward:

MadLittleMods committed:
`satisfied` typo

MadLittleMods committed:
Fix `JSONArraySome` logic

See https://github.com/matrix-org/complement/pull/796#discussion_r3318832813

MadLittleMods commented on client/client.go:
```diff
@@ -285,6 +285,39 @@ func (c *CSAPI) InviteRoom(t ct.TestLike, roomID string, userID string) *http.Re
+		serverNameStrings[i] = string(serverName)
+	}
+	query := url.Values{
+		"via": serverNameStrings,
```
According to the spec `POST /_matrix/client/v3/knock/{roomIdOrAlias}` uses the `via` query parameter, https://spec.matrix.org/v1.18/client-server-api/#post_matrixclientv3knockroomidoralias

MadLittleMods commented on match/json.go:
```diff
@@ -308,6 +308,37 @@ func JSONMapEach(wantKey string, fn func(k, v gjson.Result) error) JSON {
+		var satisifed bool = false
+		body.ForEach(func(_, val gjson.Result) bool {
+			err := fn(val)
+			satisifed = err != nil
```
Added some tests as this was a dumb mistake :white_check_mark:

anoadragon453 reviewed (approved):
Only nits below. Otherwise this LGTM!

MadLittleMods committed:
Merge branch 'main' into travis/msc4311-v2

MadLittleMods committed:
Fix `serverNames` arg docstring typo

MadLittleMods committed:
Fix markdown syntax typo in comment

See https://github.com/matrix-org/complement/pull/796#discussion_r3318960500

MadLittleMods committed:
Fix `room` plural typo

See https://github.com/matrix-org/complement/pull/796#discussion_r3319031576

MadLittleMods committed:
Clarify why run later "to allow some time for the ecosystem to adapt and support MSC4311"

See https://github.com/matrix-org/complement/pull/796#discussion_r3319050738

MadLittleMods committed:
Cross-link https://github.com/element-hq/synapse/issues/19943

MadLittleMods committed:
Fix other new typos

MadLittleMods committed:
Revert non-new typo

MadLittleMods committed:
Merge branch 'main' into travis/msc4311-v2
